src/main/java/com/metamx/emitter/core/HttpPostEmitter.java

```diff
@@ -280,8 +283,11 @@ public void run()
           for (final List<byte[]> batch : batches) {
             log.debug("Sending batch to url[%s], batch.size[%,d]", url, batch.size());
 
+            byte[] decodedBatch = config.getCompress() ? serializeAndCompressBatch(batch) : serializeBatch(batch);
```

Comment:
if there's a bug in `serializeAndCompressBatch` that prevents a specific batch from serializing properly (IOException), this has the potential to retry forever. so I suggest making the code paths in `serializeAndCompressBatch` package protected instead of `private` and checking the methods explicitly in unit tests.